File: packages/native-plugins/location/ios/Sources/LocationPlugin/LocationPlugin.swift
import Foundation
import Capacitor
import CoreLocation

/// Native iOS implementation of the ElizaLocation Capacitor plugin.
///
/// Bridges CLLocationManager to the TypeScript LocationPlugin interface, providing:
///   - getCurrentPosition (one-shot with accuracy, maxAge cache, timeout)
///   - watchPosition (continuous updates with minDistance + minInterval throttle)
///   - clearWatch (stop a running watch)
///   - checkPermissions / requestPermissions (whenInUse or always)
///   - Events: locationChange, error
@objc(ElizaLocationPlugin)
public class ElizaLocationPlugin: CAPPlugin, CAPBridgedPlugin, CLLocationManagerDelegate {
    public let identifier = "ElizaLocationPlugin"
    public let jsName = "ElizaLocation"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "getCurrentPosition", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "watchPosition", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "clearWatch", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "checkPermissions", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "requestPermissions", returnType: CAPPluginReturnPromise),
    ]

    // MARK: - State

    /// Primary manager used for permission requests and cached-location reads.
    private var locationManager: CLLocationManager!

    /// Retained manager for an in-flight one-shot location request.
    /// Must be kept alive until the delegate fires or the timeout expires.
    private var singleRequestManager: CLLocationManager?
    private var singleRequestTimer: DispatchWorkItem?

    /// Active watch sessions keyed by watch ID.
    private var watches: [String: WatchState] = [:]

    /// A pending plugin call waiting for authorization before it can proceed.
    private var pendingCall: CAPPluginCall?
    private var pendingAction: PendingAction?

    private enum PendingAction {
        case getCurrentPosition
        case watchPosition
        case requestPermissions
        case singleLocation
    }

    /// Per-watch bookkeeping.
    private struct WatchState {
        let manager: CLLocationManager
        /// Minimum interval (seconds) between emitted events. 0 = no throttle.
        let minInterval: TimeInterval
        /// Timestamp of the last emitted locationChange for this watch.
        var lastEmitted: Date?
    }

    // MARK: - Lifecycle

    public override func load() {
        locationManager = CLLocationManager()
        locationManager.delegate = self
    }

    // MARK: - getCurrentPosition

    @objc func getCurrentPosition(_ call: CAPPluginCall) {
        guard CLLocationManager.locationServicesEnabled() else {
            call.reject("Location services disabled", "POSITION_UNAVAILABLE")
            return
        }

        let status = currentAuthStatus()

        if status == .notDetermined {
            pendingCall = call
            pendingAction = .getCurrentPosition
            locationManager.requestWhenInUseAuthorization()
            return
        }

        guard status == .authorizedWhenInUse || status == .authorizedAlways else {
            call.reject("Location permission denied", "PERMISSION_DENIED")
            return
        }

        getCurrentPositionInternal(call)
    }

    private func getCurrentPositionInternal(_ call: CAPPluginCall) {
        let accuracy = call.getString("accuracy") ?? "high"
        let timeout = call.getDouble("timeout") ?? 10000
        let maxAge = call.getDouble("maxAge") ?? 0

        // Return a cached location if fresh enough (mirrors classic LocationService).
        if maxAge > 0, let cached = locationManager.location {
            let ageMs = Date().timeIntervalSince(cached.timestamp) * 1000
            if ageMs <= maxAge {
                call.resolve(buildLocationResult(from: cached, cached: true))
                return
            }
        }

        // Spin up a dedicated manager so desiredAccuracy is isolated.
        let manager = CLLocationManager()
        manager.delegate = self
        manager.desiredAccuracy = clAccuracy(from: accuracy)

        // Retain to prevent dealloc before the delegate fires.
        singleRequestManager = manager
        pendingCall = call
        pendingAction = .singleLocation

        manager.requestLocation()

        // Timeout guard.
        let timer = DispatchWorkItem { [weak self] in
            guard let self, self.pendingAction == .singleLocation else { return }
            self.cleanupSingleRequest()?.reject("Location request timed out", "TIMEOUT")
        }
        singleRequestTimer = timer
        DispatchQueue.main.asyncAfter(deadline: .now() + timeout / 1000, execute: timer)
    }

    /// Cancel in-flight single-request state and return the call (if any) for the caller to resolve/reject.
    @discardableResult
    private func cleanupSingleRequest() -> CAPPluginCall? {
        singleRequestTimer?.cancel()
        singleRequestTimer = nil
        singleRequestManager?.delegate = nil
        singleRequestManager = nil
        let call = pendingCall
        pendingCall = nil
        pendingAction = nil
        return call
    }

    // MARK: - watchPosition

    @objc func watchPosition(_ call: CAPPluginCall) {
        guard CLLocationManager.locationServicesEnabled() else {
            call.reject("Location services disabled", "POSITION_UNAVAILABLE")
            return
        }

        let status = currentAuthStatus()

        if status == .notDetermined {
            pendingCall = call
            pendingAction = .watchPosition
            locationManager.requestWhenInUseAuthorization()
            return
        }

        guard status == .authorizedWhenInUse || status == .authorizedAlways else {
            call.reject("Location permission denied", "PERMISSION_DENIED")
            return
        }

        watchPositionInternal(call)
    }

    private func watchPositionInternal(_ call: CAPPluginCall) {
        let accuracy = call.getString("accuracy") ?? "high"
        let minDistance = call.getDouble("minDistance") ?? 0
        let minInterval = call.getDouble("minInterval") ?? 0

        let watchId = UUID().uuidString
        let manager = CLLocationManager()
        manager.delegate = self
        manager.desiredAccuracy = clAccuracy(from: accuracy)
        manager.distanceFilter = minDistance > 0 ? minDistance : kCLDistanceFilterNone

        watches[watchId] = WatchState(
            manager: manager,
            minInterval: minInterval / 1000, // ms → seconds
            lastEmitted: nil
        )

        manager.startUpdatingLocation()

        call.resolve(["watchId": watchId])
    }

    // MARK: - clearWatch

    @objc func clearWatch(_ call: CAPPluginCall) {
        guard let watchId = call.getString("watchId") else {
            call.reject("Missing watchId")
            return
        }

        if let state = watches.removeValue(forKey: watchId) {
            state.manager.stopUpdatingLocation()
            state.manager.delegate = nil
        }
        call.resolve()
    }

    // MARK: - Permissions

    @objc public override func checkPermissions(_ call: CAPPluginCall) {
        call.resolve(buildPermissionResult())
    }

    @objc public override func requestPermissions(_ call: CAPPluginCall) {
        guard CLLocationManager.locationServicesEnabled() else {
            call.reject("Location services disabled", "POSITION_UNAVAILABLE")
            return
        }

        let status = currentAuthStatus()
        let level = call.getString("level") ?? "whenInUse"

        if status == .notDetermined {
            pendingCall = call
            pendingAction = .requestPermissions
            if level == "always" {
                locationManager.requestAlwaysAuthorization()
            } else {
                locationManager.requestWhenInUseAuthorization()
            }
            return
        }

        // Escalate whenInUse → always if requested (mirrors classic ensureAuthorization).
        if level == "always" && status == .authorizedWhenInUse {
            pendingCall = call
            pendingAction = .requestPermissions
            locationManager.requestAlwaysAuthorization()
            return
        }

        // Already determined — return current state.
        call.resolve(buildPermissionResult())
    }

    // MARK: - CLLocationManagerDelegate

    public func locationManager(
        _ manager: CLLocationManager,
        didUpdateLocations locations: [CLLocation]
    ) {
        guard let location = locations.last else { return }

        // One-shot request?
        if pendingAction == .singleLocation, manager === singleRequestManager {
            cleanupSingleRequest()?.resolve(buildLocationResult(from: location, cached: false))
            return
        }

        // Watch update — find the matching watch and apply minInterval throttle.
        for (watchId, var state) in watches where state.manager === manager {
            if state.minInterval > 0, let last = state.lastEmitted,
               Date().timeIntervalSince(last) < state.minInterval
            {
                return // throttled
            }
            state.lastEmitted = Date()
            watches[watchId] = state
            notifyListeners("locationChange", data: buildLocationResult(from: location, cached: false))
            return
        }
    }

    public func locationManager(
        _ manager: CLLocationManager,
        didFailWithError error: Error
    ) {
        // One-shot request?
        if pendingAction == .singleLocation, manager === singleRequestManager {
            cleanupSingleRequest()?.reject(
                "Location error: \(error.localizedDescription)", "POSITION_UNAVAILABLE"
            )
            return
        }

        // Watch error — emit event.
        for (_, state) in watches where state.manager === manager {
            notifyListeners("error", data: [
                "code": "POSITION_UNAVAILABLE",
                "message": error.localizedDescription,
            ])
            return
        }
    }

    public func locationManager(
        _ manager: CLLocationManager,
        didChangeAuthorization status: CLAuthorizationStatus
    ) {
        guard let call = pendingCall, let action = pendingAction else { return }

        // Still waiting for user to decide.
        if status == .notDetermined { return }

        // Clear pending state *before* calling internal methods — they may set new values.
        pendingCall = nil
        pendingAction = nil

        switch action {
        case .getCurrentPosition:
            if status == .authorizedWhenInUse || status == .authorizedAlways {
                getCurrentPositionInternal(call)
            } else {
                call.reject("Location permission denied", "PERMISSION_DENIED")
            }

        case .watchPosition:
            if status == .authorizedWhenInUse || status == .authorizedAlways {
                watchPositionInternal(call)
            } else {
                call.reject("Location permission denied", "PERMISSION_DENIED")
            }

        case .requestPermissions:
            call.resolve(buildPermissionResult())

        case .singleLocation:
            // Shouldn't happen — singleLocation is set after auth is granted.
            break
        }
    }

    // MARK: - Helpers

    private func currentAuthStatus() -> CLAuthorizationStatus {
        CLLocationManager.authorizationStatus()
    }

    /// Map the TypeScript LocationAccuracy string to a CLLocationAccuracy constant.
    private func clAccuracy(from accuracy: String) -> CLLocationAccuracy {
        switch accuracy {
        case "best":    return kCLLocationAccuracyBest
        case "high":    return kCLLocationAccuracyNearestTenMeters
        case "medium":  return kCLLocationAccuracyHundredMeters
        case "low":     return kCLLocationAccuracyKilometer
        case "passive": return kCLLocationAccuracyThreeKilometers
        default:        return kCLLocationAccuracyNearestTenMeters
        }
    }

    /// Build the permission result matching LocationPermissionStatus in definitions.ts.
    private func buildPermissionResult() -> JSObject {
        let status = currentAuthStatus()
        let location: String
        let background: String

        switch status {
        case .authorizedAlways:
            location = "granted"
            background = "granted"
        case .authorizedWhenInUse:
            location = "granted"
            background = "prompt"
        case .denied, .restricted:
            location = "denied"
            background = "denied"
        default:
            location = "prompt"
            background = "prompt"
        }

        return ["location": location, "background": background]
    }

    /// Build a LocationResult matching the TypeScript interface:
    /// `{ coords: LocationCoordinates, cached: boolean }`
    private func buildLocationResult(from location: CLLocation, cached: Bool) -> JSObject {
        var coords: JSObject = [
            "latitude": location.coordinate.latitude,
            "longitude": location.coordinate.longitude,
            "accuracy": location.horizontalAccuracy,
            "timestamp": location.timestamp.timeIntervalSince1970 * 1000,
        ]

        // Altitude data is valid when verticalAccuracy >= 0.
        if location.verticalAccuracy >= 0 {
            coords["altitude"] = location.altitude
            coords["altitudeAccuracy"] = location.verticalAccuracy
        }
        if location.speed >= 0 {
            coords["speed"] = location.speed
        }
        if location.course >= 0 {
            coords["heading"] = location.course
        }

        return ["coords": coords, "cached": cached]
    }
}

File: packages/native-plugins/location/android/src/main/java/ai/eliza/plugins/location/LocationPlugin.kt
package ai.eliza.plugins.location

import android.Manifest
import android.content.pm.PackageManager
import android.location.Location
import android.os.Build
import android.os.Looper
import androidx.core.content.ContextCompat
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.getcapacitor.annotation.Permission
import com.getcapacitor.annotation.PermissionCallback
import com.google.android.gms.location.*
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

/**
 * ElizaLocation Capacitor Plugin
 *
 * Provides location services using Google Play Services FusedLocationProviderClient.
 * Supports one-shot position, continuous watching, maxAge caching, and background location.
 */
@CapacitorPlugin(
    name = "ElizaLocation",
    permissions = [
        Permission(alias = "location", strings = [
            Manifest.permission.ACCESS_FINE_LOCATION,
            Manifest.permission.ACCESS_COARSE_LOCATION
        ]),
        Permission(alias = "background", strings = [
            Manifest.permission.ACCESS_BACKGROUND_LOCATION
        ])
    ]
)
class LocationPlugin : Plugin() {

    private var fusedLocationClient: FusedLocationProviderClient? = null
    private val watches = ConcurrentHashMap<String, LocationCallback>()
    private var pendingCall: PluginCall? = null
    private var pendingAction: String? = null

    // Cache the last known location for maxAge support
    private var lastKnownLocation: Location? = null

    override fun load() {
        super.load()
        fusedLocationClient = LocationServices.getFusedLocationProviderClient(activity)
    }

    // ── getCurrentPosition ──────────────────────────────────────────────

    @PluginMethod
    fun getCurrentPosition(call: PluginCall) {
        if (!hasRequiredPermissions()) {
            pendingCall = call
            pendingAction = "getCurrentPosition"
            requestAllPermissions(call, "handlePermissionResult")
            return
        }
        getCurrentPositionInternal(call)
    }

    private fun getCurrentPositionInternal(call: PluginCall) {
        val accuracy = call.getString("accuracy") ?: "high"
        val timeout = call.getDouble("timeout") ?: 10000.0
        val maxAge = call.getDouble("maxAge") ?: 0.0
        val priority = mapAccuracyToPriority(accuracy)

        // maxAge > 0: try returning cached location if fresh enough (mirrors classic bestLastKnown)
        if (maxAge > 0) {
            try {
                fusedLocationClient?.lastLocation?.addOnSuccessListener { cached ->
                    if (cached != null) {
                        val age = System.currentTimeMillis() - cached.time
                        if (age <= maxAge.toLong()) {
                            lastKnownLocation = cached
                            call.resolve(buildLocationResult(cached, cached = true))
                            return@addOnSuccessListener
                        }
                    }
                    // Cache miss — fall through to a fresh fix
                    requestFreshLocation(call, priority, timeout, maxAge)
                }?.addOnFailureListener {
                    requestFreshLocation(call, priority, timeout, maxAge)
                }
                return
            } catch (_: SecurityException) {
                // Permission lost between check and call — fall through
            }
        }

        requestFreshLocation(call, priority, timeout, maxAge)
    }

    /** Request a fresh location using CurrentLocationRequest. */
    private fun requestFreshLocation(call: PluginCall, priority: Int, timeout: Double, maxAge: Double) {
        val request = CurrentLocationRequest.Builder()
            .setPriority(priority)
            .setMaxUpdateAgeMillis(maxAge.toLong())
            .setDurationMillis(timeout.toLong())
            .build()

        try {
            fusedLocationClient?.getCurrentLocation(request, null)
                ?.addOnSuccessListener { location ->
                    if (location != null) {
                        lastKnownLocation = location
                        call.resolve(buildLocationResult(location, cached = false))
                    } else {
                        val err = buildErrorEvent("POSITION_UNAVAILABLE", "Unable to get location")
                        notifyListeners("error", err)
                        call.reject("Unable to get location")
                    }
                }
                ?.addOnFailureListener { e ->
                    val code = if (e is SecurityException) "PERMISSION_DENIED" else "POSITION_UNAVAILABLE"
                    val err = buildErrorEvent(code, "Location error: ${e.message}")
                    notifyListeners("error", err)
                    call.reject("Location error: ${e.message}")
                }
        } catch (e: SecurityException) {
            val err = buildErrorEvent("PERMISSION_DENIED", "Location permission required")
            notifyListeners("error", err)
            call.reject("Location permission required")
        }
    }

    // ── watchPosition ───────────────────────────────────────────────────

    @PluginMethod
    fun watchPosition(call: PluginCall) {
        if (!hasRequiredPermissions()) {
            pendingCall = call
            pendingAction = "watchPosition"
            requestAllPermissions(call, "handlePermissionResult")
            return
        }
        watchPositionInternal(call)
    }

    private fun watchPositionInternal(call: PluginCall) {
        val accuracy = call.getString("accuracy") ?: "high"
        val minInterval = call.getDouble("minInterval") ?: 0.0
        val minDistance = call.getDouble("minDistance") ?: 0.0
        val priority = mapAccuracyToPriority(accuracy)

        val watchId = UUID.randomUUID().toString()

        val request = LocationRequest.Builder(priority, minInterval.toLong())
            .setMinUpdateDistanceMeters(minDistance.toFloat())
            .build()

        val callback = object : LocationCallback() {
            override fun onLocationResult(result: LocationResult) {
                for (location in result.locations) {
                    lastKnownLocation = location
                    notifyListeners("locationChange", buildLocationResult(location, cached = false))
                }
            }

            override fun onLocationAvailability(availability: LocationAvailability) {
                if (!availability.isLocationAvailable) {
                    notifyListeners("error", buildErrorEvent(
                        "POSITION_UNAVAILABLE",
                        "Location services became unavailable"
                    ))
                }
            }
        }

        try {
            fusedLocationClient?.requestLocationUpdates(
                request,
                callback,
                Looper.getMainLooper()
            )

            watches[watchId] = callback
            call.resolve(JSObject().apply {
                put("watchId", watchId)
            })
        } catch (e: SecurityException) {
            notifyListeners("error", buildErrorEvent("PERMISSION_DENIED", "Location permission required"))
            call.reject("Location permission required")
        }
    }

    // ── clearWatch ──────────────────────────────────────────────────────

    @PluginMethod
    fun clearWatch(call: PluginCall) {
        val watchId = call.getString("watchId")
        if (watchId == null) {
            call.reject("Missing watchId")
            return
        }

        val callback = watches.remove(watchId)
        if (callback != null) {
            fusedLocationClient?.removeLocationUpdates(callback)
        }
        call.resolve()
    }

    // ── Permissions ─────────────────────────────────────────────────────

    @PluginMethod
    override fun checkPermissions(call: PluginCall) {
        call.resolve(buildPermissionResult())
    }

    @PluginMethod
    override fun requestPermissions(call: PluginCall) {
        if (hasRequiredPermissions()) {
            call.resolve(buildPermissionResult())
            return
        }
        pendingAction = "requestPermissions"
        requestAllPermissions(call, "handlePermissionResult")
    }

    @PermissionCallback
    private fun handlePermissionResult(call: PluginCall) {
        if (hasRequiredPermissions()) {
            when (pendingAction) {
                "getCurrentPosition" -> {
                    pendingAction = null
                    pendingCall = null
                    getCurrentPositionInternal(call)
                }
                "watchPosition" -> {
                    pendingAction = null
                    pendingCall = null
                    watchPositionInternal(call)
                }
                else -> {
                    pendingAction = null
                    pendingCall = null
                    call.resolve(buildPermissionResult())
                }
            }
        } else {
            pendingAction = null
            pendingCall = null
            notifyListeners("error", buildErrorEvent("PERMISSION_DENIED", "Location permission denied"))
            call.reject("Location permission denied")
        }
    }

    // ── Helpers ──────────────────────────────────────────────────────────

    override fun hasRequiredPermissions(): Boolean {
        return getPermissionState("location") == com.getcapacitor.PermissionState.GRANTED
    }

    /** Map accuracy string from JS to Play Services Priority constant. */
    private fun mapAccuracyToPriority(accuracy: String): Int {
        return when (accuracy) {
            "best" -> Priority.PRIORITY_HIGH_ACCURACY
            "high" -> Priority.PRIORITY_HIGH_ACCURACY
            "medium" -> Priority.PRIORITY_BALANCED_POWER_ACCURACY
            "low" -> Priority.PRIORITY_LOW_POWER
            "passive" -> Priority.PRIORITY_PASSIVE
            else -> Priority.PRIORITY_HIGH_ACCURACY
        }
    }

    private fun buildPermissionResult(): JSObject {
        val locationState = getPermissionState("location")
        val locationStatus = when (locationState) {
            com.getcapacitor.PermissionState.GRANTED -> "granted"
            com.getcapacitor.PermissionState.DENIED -> "denied"
            else -> "prompt"
        }

        val result = JSObject().apply {
            put("location", locationStatus)
        }

        // Background location is a separate permission on Android 10+
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            val bgGranted = ContextCompat.checkSelfPermission(
                context, Manifest.permission.ACCESS_BACKGROUND_LOCATION
            ) == PackageManager.PERMISSION_GRANTED

            val bgStatus = when {
                bgGranted -> "granted"
                // If foreground isn't granted, background is implicitly denied
                locationStatus != "granted" -> "denied"
                else -> "prompt"
            }
            result.put("background", bgStatus)
        } else {
            // Pre-Q: background is granted with foreground
            result.put("background", locationStatus)
        }

        return result
    }

    private fun buildLocationResult(location: Location, cached: Boolean): JSObject {
        val coords = JSObject().apply {
            put("latitude", location.latitude)
            put("longitude", location.longitude)
            if (location.hasAltitude()) {
                put("altitude", location.altitude)
            }
            put("accuracy", location.accuracy.toDouble())
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O && location.hasVerticalAccuracy()) {
                put("altitudeAccuracy", location.verticalAccuracyMeters.toDouble())
            }
            if (location.hasSpeed()) {
                put("speed", location.speed.toDouble())
            }
            if (location.hasBearing()) {
                put("heading", location.bearing.toDouble())
            }
            put("timestamp", location.time)
        }

        return JSObject().apply {
            put("coords", coords)
            put("cached", cached)
        }
    }

    private fun buildErrorEvent(code: String, message: String): JSObject {
        return JSObject().apply {
            put("code", code)
            put("message", message)
        }
    }

    // ── Lifecycle ───────────────────────────────────────────────────────

    override fun handleOnDestroy() {
        super.handleOnDestroy()
        for ((_, callback) in watches) {
            fusedLocationClient?.removeLocationUpdates(callback)
        }
        watches.clear()
    }
}
